### PR TITLE
feat: Integrate fee token

### PR DESCRIPTION
**Pull request by deluca-mike**
### Replace app chain native token with fee token wrapper system across settlement chain gateway, payer registry, and distribution manager contracts
This pull request introduces a fee token wrapper system that replaces direct usage of app chain native tokens throughout the codebase. The changes include:

• Replace `appChainNativeToken` with `feeToken` and `underlyingFeeToken` across [SettlementChainGateway.sol](https://github.com/xmtp/smart-contracts/pull/90/files#diff-78163df21c1597c48e2b008d097dfd46c2771fa1184cf52353b5dd33dd8d4c2f), [PayerRegistry.sol](https://github.com/xmtp/smart-contracts/pull/90/files#diff-9a732bc06667d4c6d432fed365fe1b3b27f0a8937ee5df21bc6c02cd506d529f), and [DistributionManager.sol](https://github.com/xmtp/smart-contracts/pull/90/files#diff-13865dcd192ab96ef2fe95fcf550b7957883271fe85745961659db93987ccbdd)
• Add new deposit and withdrawal methods that support both fee tokens and underlying tokens, including `depositFromUnderlying`, `depositFromUnderlyingWithPermit`, `withdrawIntoUnderlying`, and `finalizeWithdrawalIntoUnderlying`
• Implement pause functionality across gateway and distribution manager contracts with `updatePauseStatus` methods and `whenNotPaused` modifiers
• Add withdrawal functionality to [AppChainGateway.sol](https://github.com/xmtp/smart-contracts/pull/90/files#diff-4591f77b348f45bb9cedef8d67041599de2961ce5d7af5f2ac9dfcb375a926da) with `withdraw` and `withdrawIntoUnderlying` methods that use Arbitrum system contract
• Replace inbox address parameters with chain ID parameters in settlement chain gateway functions
• Update deployment scripts to deploy fee token implementations and proxies
• Add new [FeeTokenDeployer.sol](https://github.com/xmtp/smart-contracts/pull/90/files#diff-6017bafe2f16000fd43eef0d5346d9a9c71b75e7a0aa83887d4a533e12780f85) for fee token contract deployment
• Modify test files to use direct string comparison for parameter keys instead of hash comparison

#### 📍Where to Start
Start with the `SettlementChainGateway` constructor and `deposit` function in [SettlementChainGateway.sol](https://github.com/xmtp/smart-contracts/pull/90/files#diff-78163df21c1597c48e2b008d097dfd46c2771fa1184cf52353b5dd33dd8d4c2f) to understand how the fee token system replaces the previous app chain native token approach.

----

_[Macroscope](https://app.macroscope.com) summarized 7f9a4b8._

## Summary by CodeRabbit

* **New Features**
  * Introduced a flexible fee token system with support for underlying tokens, replacing the previous native token approach across all chains.
  * Added deposit and withdrawal options for both fee tokens and underlying tokens, including permit-based and multi-chain operations.
  * Enabled contract pausing and unpausing, with new functions to update and query pause status.
  * Enhanced withdrawal functionality, allowing unwrapping to underlying tokens and L2-to-L1 messaging integration.
  * Expanded event logging and error reporting for deposits, withdrawals, and parameter updates.

* **Bug Fixes**
  * Improved validation and error handling for zero addresses, unsupported chain IDs, and redundant state updates.

* **Refactor**
  * Updated deployment scripts, interfaces, and tests to align with the new fee token architecture.
  * Standardized naming conventions and parameter handling for clarity and future extensibility.

* **Tests**
  * Added comprehensive tests for new deposit, withdrawal, pause, and migration flows, including edge cases and failure scenarios.

* **Chores**
  * Updated configuration files and removed obsolete environment files to reflect the new token system and deployment structure.

### CODE REVIEW

File: script/Parameters.s.sol [hunk 1]
```diff
@@ -119,26 +123,22 @@ contract ParameterScripts is Script {
         keys_[2] = _IDENTITY_UPDATE_BROADCASTER_MIN_PAYLOAD_SIZE_KEY;
         keys_[3] = _IDENTITY_UPDATE_BROADCASTER_MAX_PAYLOAD_SIZE_KEY;
 
-        address[] memory inboxes_ = new address[](1);
-        inboxes_[0] = _deploymentData.settlementChainInboxToAppchain;
-
-        vm.startBroadcast(_privateKey);
+        uint256 gasLimit_ = _TX_STIPEND + (_GAS_PER_BRIDGED_KEY * keys_.length);
+        uint256 cost_ = (_APP_CHAIN_GAS_PRICE * gasLimit_) / 1e18;
```
[the remark this answers is not in the document]
Comment (deluca-mike) [reply script/Parameters.s.sol:127]: Good catch, but actually the error here is that it assumes the FeeToken has no decimals. The fix should be `uint256 cost_ = ((_APP_CHAIN_GAS_PRICE * gasLimit_) * 1e6) / 1e18;` since we know the fee token has 6 decimals and that gas on an EVM (app chain) has 18 decimals.

File: script/utils/FeeTokenBridging.s.sol [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+contract MockArbSys {
+    function withdrawEth(address recipient_) external payable {}
+}
+
+contract FeeTokenBridging is Script {
+    address internal constant _APPCHAIN_NATIVE_TOKEN = 0x036CbD53842c5426634e7929541eC2318f3dCF7e;
+    address internal constant _SETTLEMENT_CHAIN_INBOX_TO_APPCHAIN = 0xd06d8E471F0EeB1bb516303EdE399d004Acb1615;
+    address internal constant _SETTLEMENT_CHAIN_OUTBOX_FROM_APPCHAIN = 0xE401acFc336524955152b7260df47c3E23e4A54A;
+    address internal constant _ARB_SYS = 0x0000000000000000000000000000000000000064; // address(100)
+    address internal constant _NODE_INTERFACE = 0x00000000000000000000000000000000000000C8; // address(200)
```
[the remark this answers is not in the document]
Comment (deluca-mike) [reply script/utils/FeeTokenBridging.s.sol:25]: Great, but this whole file was included by mistake. Will be removed.

File: script/utils/FeeTokenBridging.s.sol [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    function withdraw(uint256 amount_) external {
+        vm.createSelectFork(vm.envString("XMTP_TESTNET_RPC_URL"));
+
+        vm.etch(_ARB_SYS, type(MockArbSys).runtimeCode);
+
+        vm.startBroadcast(_privateKey);
+        (bool success_, ) = _ARB_SYS.call{ value: amount_ }(abi.encodeWithSignature("withdrawEth(address)", _account));
+        if (!success_) revert WithdrawFailed();
+        vm.stopBroadcast();
+    }
```
[the remark this answers is not in the document]
Comment (deluca-mike) [reply script/utils/FeeTokenBridging.s.sol:77]: Great, but this whole file was included by mistake. Will be removed.

File: test/integration/Deploy.t.sol [hunk 4]
```diff
@@ -1200,7 +1312,7 @@
     /* ============ Expected Proxy Getters ============ */
 
     function _expectedGatewayProxy() internal returns (address expectedGatewayProxy_) {
-        vm.selectFork(_appChainForkId);
-        return _appChainFactory.computeProxyAddress(_deployer, _GATEWAY_PROXY_SALT);
+        vm.selectFork(_settlementChainForkId);
+        return _settlementChainFactory.computeProxyAddress(_deployer, _GATEWAY_PROXY_SALT);
     }
```
[the remark this answers is not in the document]
Comment (deluca-mike) [reply test/integration/Deploy.t.sol:1317]: This is intended. The process is done on the settlement chain first, and the factories and gateways have the same address on both chains.

**Review by fbac — APPROVED** [review]
